File: emarsys-sdk/src/androidTest/java/com/emarsys/config/EmarsysConfigTest.java
package com.emarsys.config;

import android.app.Application;

import com.emarsys.core.api.experimental.FlipperFeature;
import com.emarsys.mobileengage.api.NotificationEventHandler;
import com.emarsys.mobileengage.api.event.EventHandler;
import com.emarsys.testUtil.InstrumentationRegistry;
import com.emarsys.testUtil.TimeoutUtils;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;

import static junit.framework.TestCase.assertTrue;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.fail;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;

public class EmarsysConfigTest {
    private String APP_ID = "appID";
    private int CONTACT_FIELD_ID = 567;
    private String MERCHANT_ID = "MERCHANT_ID";
    private Application application;
    private EventHandler defaultInAppEventHandler;
    private EventHandler defaultNotificationEventHandler;
    private FlipperFeature[] features;
    private boolean automaticPushTokenSending;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    public void init() {
        automaticPushTokenSending = true;
        application = (Application) InstrumentationRegistry.getTargetContext().getApplicationContext();
        defaultInAppEventHandler = mock(EventHandler.class);
        defaultNotificationEventHandler = mock(EventHandler.class);
        features = new FlipperFeature[]{
                mock(FlipperFeature.class),
                mock(FlipperFeature.class)
        };
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_applicationShouldNotBeNull() {
        new EmarsysConfig(
                null,
                APP_ID,
                CONTACT_FIELD_ID,
                MERCHANT_ID,
                defaultInAppEventHandler,
                defaultNotificationEventHandler,
                features,
                automaticPushTokenSending);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_contactFieldIdShouldNotBeNull() {
        new EmarsysConfig(
                application,
                APP_ID,
                null,
                MERCHANT_ID,
                defaultInAppEventHandler,
                defaultNotificationEventHandler,
                features,
                automaticPushTokenSending);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_featuresShouldNotBeNull() {
        new EmarsysConfig(
                application,
                APP_ID,
                CONTACT_FIELD_ID,
                MERCHANT_ID,
                defaultInAppEventHandler,
                defaultNotificationEventHandler,
                null,
                automaticPushTokenSending);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_featuresList_shouldNotContainNullElements() {
        new EmarsysConfig(
                application,
                APP_ID,
                CONTACT_FIELD_ID,
                MERCHANT_ID,
                defaultInAppEventHandler,
                defaultNotificationEventHandler,
                new FlipperFeature[]{mock(FlipperFeature.class), null},
                automaticPushTokenSending);
    }

    @Test
    public void testBuilder_withAllArguments() {
        EmarsysConfig expected = new EmarsysConfig(
                application,
                APP_ID,
                CONTACT_FIELD_ID,
                MERCHANT_ID,
                defaultInAppEventHandler,
                defaultNotificationEventHandler,
                features,
                automaticPushTokenSending);

        EmarsysConfig result = new EmarsysConfig.Builder()
                .application(application)
                .mobileEngageApplicationCode(APP_ID)
                .contactFieldId(CONTACT_FIELD_ID)
                .predictMerchantId(MERCHANT_ID)
                .enableExperimentalFeatures(features)
                .inAppEventHandler(any(com.emarsys.mobileengage.api.EventHandler.class))
                .notificationEventHandler(any(NotificationEventHandler.class))
                .build();

        assertEquals(expected.getApplication(), result.getApplication());
        assertEquals(expected.getContactFieldId(), result.getContactFieldId());
        assertEquals(expected.getExperimentalFeatures(), result.getExperimentalFeatures());
        assertEquals(expected.getMobileEngageApplicationCode(), result.getMobileEngageApplicationCode());
        assertEquals(expected.getPredictMerchantId(), result.getPredictMerchantId());
        Assert.assertTrue(result.getInAppEventHandler().getClass().isInstance(expected.getInAppEventHandler()));
        Assert.assertTrue(result.getNotificationEventHandler().getClass().isInstance(expected.getNotificationEventHandler()));
    }

    @Test
    public void testBuilder_withRequiredArguments() {
        EmarsysConfig expected = new EmarsysConfig(
                application,
                APP_ID,
                CONTACT_FIELD_ID,
                MERCHANT_ID,
                null,
                null,
                new FlipperFeature[]{},
                automaticPushTokenSending);

        EmarsysConfig result = new EmarsysConfig.Builder()
                .application(application)
                .mobileEngageApplicationCode(APP_ID)
                .contactFieldId(CONTACT_FIELD_ID)
                .predictMerchantId(MERCHANT_ID)
                .build();

        assertEquals(expected, result);
    }

    @Test
    public void testBuilder_whenInAppMessagingFlipperIsOff_defaultInAppMessageHandlerIsNotRequired() {
        try {
            new EmarsysConfig.Builder()
                    .application(application)
                    .mobileEngageApplicationCode(APP_ID)
                    .contactFieldId(CONTACT_FIELD_ID)
                    .predictMerchantId(MERCHANT_ID)
                    .build();
        } catch (IllegalArgumentException e) {
            fail("Should not fail with: " + e.getMessage());
        }
    }

    @Test
    public void testBuilder_automaticPushTokenSending_whenDisabled() {
        EmarsysConfig config = new EmarsysConfig.Builder()
                .application(application)
                .mobileEngageApplicationCode(APP_ID)
                .contactFieldId(CONTACT_FIELD_ID)
                .predictMerchantId(MERCHANT_ID)
                .disableAutomaticPushTokenSending()
                .build();
        assertFalse(config.isAutomaticPushTokenSendingEnabled());
    }

    @Test
    public void testBuilder_automaticPushTokenSending_default() {
        EmarsysConfig config = new EmarsysConfig.Builder()
                .application(application)
                .mobileEngageApplicationCode(APP_ID)
                .contactFieldId(CONTACT_FIELD_ID)
                .predictMerchantId(MERCHANT_ID)
                .build();
        assertTrue(config.isAutomaticPushTokenSendingEnabled());
    }

    @Test(expected = IllegalArgumentException.class)
    public void
    testBuilder_from_shouldNotAcceptNull() {
        new EmarsysConfig.Builder().from(null);
    }

    @Test
    public void testBuilder_from() {
        EmarsysConfig expected = new EmarsysConfig(
                application,
                APP_ID,
                CONTACT_FIELD_ID,
                MERCHANT_ID,
                defaultInAppEventHandler,
                defaultNotificationEventHandler,
                features,
                automaticPushTokenSending);

        EmarsysConfig result = new EmarsysConfig.Builder()
                .from(expected)
                .build();

        assertEquals(expected.getApplication(), result.getApplication());
        assertEquals(expected.getContactFieldId(), result.getContactFieldId());
        assertEquals(expected.getExperimentalFeatures(), result.getExperimentalFeatures());
        assertEquals(expected.getMobileEngageApplicationCode(), result.getMobileEngageApplicationCode());
        assertEquals(expected.getPredictMerchantId(), result.getPredictMerchantId());
        Assert.assertTrue(result.getInAppEventHandler().getClass().isInstance(expected.getInAppEventHandler()));
        Assert.assertTrue(result.getNotificationEventHandler().getClass().isInstance(expected.getNotificationEventHandler()));
    }
}

File: emarsys-sdk/src/androidTest/java/com/emarsys/config/EmarsysConfigTest.kt
package com.emarsys.config


import android.app.Application
import com.emarsys.common.feature.InnerFeature
import com.emarsys.core.api.experimental.FlipperFeature
import com.emarsys.core.endpoint.ServiceEndpointProvider
import com.emarsys.di.FakeDependencyContainer
import com.emarsys.di.setupEmarsysComponent
import com.emarsys.mobileengage.api.event.EventHandler
import com.emarsys.testUtil.AnnotationSpec
import com.emarsys.testUtil.FeatureTestUtils
import com.emarsys.testUtil.InstrumentationRegistry.Companion.getTargetContext
import com.emarsys.testUtil.IntegrationTestUtils
import io.kotest.assertions.fail
import io.kotest.matchers.shouldBe
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock

class EmarsysConfigTest : AnnotationSpec() {
    companion object {
        private const val APP_ID = "appID"
        private const val MERCHANT_ID = "MERCHANT_ID"
        private const val SHARED_SECRET = "testSecret"
        private const val CLIENT_HOST = "https://me-client.eservice.emarsys.net"
        private const val EVENT_HOST = "https://mobile-events.eservice.emarsys.net"
        private const val INBOX_HOST = "https://me-inbox.eservice.emarsys.net/v3"
        private val SHARED_PACKAGE_NAMES = listOf("package1", "package2")
    }

    private lateinit var application: Application
    private lateinit var defaultInAppEventHandler: EventHandler
    private lateinit var defaultNotificationEventHandler: EventHandler
    private lateinit var features: Array<FlipperFeature>
    private var automaticPushTokenSending = false


    @Before
    fun setUp() {
        FeatureTestUtils.resetFeatures()
        val mockClientServiceProvider: ServiceEndpointProvider = mock {
            on { provideEndpointHost() } doReturn CLIENT_HOST
        }
        val mockEventServiceProvider: ServiceEndpointProvider = mock {
            on { provideEndpointHost() } doReturn EVENT_HOST
        }
        val mockMessageInboxServiceProvider: ServiceEndpointProvider = mock {
            on { provideEndpointHost() } doReturn INBOX_HOST
        }
        val dependencyContainer = FakeDependencyContainer(
            clientServiceEndpointProvider = mockClientServiceProvider,
            eventServiceEndpointProvider = mockEventServiceProvider,
            messageInboxServiceProvider = mockMessageInboxServiceProvider
        )

        setupEmarsysComponent(dependencyContainer)


        automaticPushTokenSending = true
        application = getTargetContext().applicationContext as Application
        defaultInAppEventHandler = mock()
        defaultNotificationEventHandler = mock()
        features = arrayOf(
            mock(),
            mock()
        )
    }

    @After
    fun tearDown() {
        IntegrationTestUtils.tearDownEmarsys(application)
    }

    @Test
    fun testBuilder_withAllArguments() {
        val expected = EmarsysConfig(
            application,
            APP_ID,
            MERCHANT_ID,
            listOf(*features),
            automaticPushTokenSending,
            SHARED_PACKAGE_NAMES,
            SHARED_SECRET,
            true
        )
        val result = EmarsysConfig.Builder()
            .application(application)
            .applicationCode(APP_ID)
            .merchantId(MERCHANT_ID)
            .enableExperimentalFeatures(*features)
            .sharedSecret("testSecret")
            .sharedPackageNames(SHARED_PACKAGE_NAMES)
            .enableVerboseConsoleLogging()
            .build()
        result.application shouldBe expected.application
        result.experimentalFeatures shouldBe expected.experimentalFeatures
        result.applicationCode shouldBe expected.applicationCode
        result.merchantId shouldBe expected.merchantId
        result.sharedSecret shouldBe expected.sharedSecret
        result.sharedPackageNames shouldBe expected.sharedPackageNames
        result.verboseConsoleLoggingEnabled shouldBe expected.verboseConsoleLoggingEnabled
    }

    @Test
    fun testBuilder_withRequiredArguments() {
        val expected = EmarsysConfig(
            application,
            APP_ID,
            MERCHANT_ID,
            emptyList(),
            automaticPushTokenSending,
            null,
            null,
            false
        )
        val result = EmarsysConfig.Builder()
            .application(application)
            .applicationCode(APP_ID)
            .merchantId(MERCHANT_ID)
            .build()

        result shouldBe expected
    }

    @Test
    fun testBuilder_whenInAppMessagingFlipperIsOff_defaultInAppMessageHandlerIsNotRequired() {
        try {
            EmarsysConfig.Builder()
                .application(application)
                .applicationCode(APP_ID)
                .merchantId(MERCHANT_ID)
                .build()
        } catch (e: IllegalArgumentException) {
            fail("Should not fail with: ${e.message}")
        }
    }

    @Test
    fun testBuilder_automaticPushTokenSending_whenDisabled() {
        val config = EmarsysConfig.Builder()
            .application(application)
            .applicationCode(APP_ID)
            .merchantId(MERCHANT_ID)
            .disableAutomaticPushTokenSending()
            .build()
        config.automaticPushTokenSendingEnabled shouldBe false
    }

    @Test
    fun testBuilder_automaticPushTokenSending_default() {
        val config = EmarsysConfig.Builder()
            .application(application)
            .applicationCode(APP_ID)
            .merchantId(MERCHANT_ID)
            .build()

        config.automaticPushTokenSendingEnabled shouldBe true
    }

    @Test
    fun testBuilder_verboseConsoleLoggingEnabled() {
        val config = EmarsysConfig.Builder()
            .application(application)
            .applicationCode(APP_ID)
            .merchantId(MERCHANT_ID)
            .enableVerboseConsoleLogging()
            .build()

        config.verboseConsoleLoggingEnabled shouldBe true
    }

    @Test
    fun testBuilder_verboseConsoleLoggingDisabled_byDefault() {
        val config = EmarsysConfig.Builder()
            .application(application)
            .applicationCode(APP_ID)
            .merchantId(MERCHANT_ID)
            .build()

        config.verboseConsoleLoggingEnabled shouldBe false
    }

    @Test
    fun testEmarsysConfig_defaultParameters() {
        val config = EmarsysConfig(
            application = application
        )

        config.application shouldBe application
        config.applicationCode shouldBe null
        config.automaticPushTokenSendingEnabled shouldBe true
        config.verboseConsoleLoggingEnabled shouldBe false
        config.experimentalFeatures shouldBe listOf()
        config.merchantId shouldBe null
        config.sharedSecret shouldBe null
        config.sharedPackageNames shouldBe null
    }

    @Test
    fun testEmarsysConfig_allParameters() {
        val config = EmarsysConfig(
            application = application,
            applicationCode = APP_ID,
            merchantId = MERCHANT_ID,
            experimentalFeatures = listOf(InnerFeature.EVENT_SERVICE_V4),
            automaticPushTokenSendingEnabled = false,
            sharedPackageNames = SHARED_PACKAGE_NAMES,
            sharedSecret = SHARED_SECRET,
            verboseConsoleLoggingEnabled = true
        )

        config.application shouldBe application
        config.applicationCode shouldBe APP_ID
        config.automaticPushTokenSendingEnabled shouldBe false
        config.verboseConsoleLoggingEnabled shouldBe true
        config.experimentalFeatures shouldBe listOf(InnerFeature.EVENT_SERVICE_V4)
        config.merchantId shouldBe MERCHANT_ID
        config.sharedSecret shouldBe SHARED_SECRET
        config.sharedPackageNames shouldBe SHARED_PACKAGE_NAMES
    }

    @Test
    fun testBuilder_from() {
        val expected = EmarsysConfig(
            application,
            APP_ID,
            MERCHANT_ID,
            listOf(*features),
            automaticPushTokenSending,
            SHARED_PACKAGE_NAMES,
            SHARED_SECRET,
            false
        )
        val result = EmarsysConfig.Builder()
            .from(expected)
            .build()
        result.application shouldBe expected.application
        result.experimentalFeatures shouldBe expected.experimentalFeatures
        result.applicationCode shouldBe expected.applicationCode
        result.merchantId shouldBe expected.merchantId
        result.sharedSecret shouldBe expected.sharedSecret
        result.sharedPackageNames shouldBe expected.sharedPackageNames
        result.verboseConsoleLoggingEnabled shouldBe false
    }
}